### ISSUE TITLE
Bug in poly

### ISSUE THREAD

**Issue by abhinav28071999** [sympy/sympy#18555]
`from sympy import Poly`
`from sympy.abc import x`
`f = Poly(x**3 + 2*x - 1, x) `
`f.subs(x, Matrix([1]))`

gives `Poly(Matrix([[1]])**3 + 2*Matrix([[1]]) - 1, Matrix([[1]]), domain='ZZ')`

**Comment by sylee957** [comment]
I don't know how `subs` works for poly, but it may not be a good way to substitute non-coefficients in `Poly` with something else, unlike expr.

**Comment by Jonty16117** [comment]
I would like to work on this issue

**Comment by Jonty16117** [comment]
> I don't know how `subs` works for poly, but it may not be a good way to substitute non-coefficients in `Poly` with something else, unlike expr.

But i think, Matrix[[1]] should be just equal to 1

**Comment by jksuom** [comment]
> Poly(Matrix([[1]])**3 + 2*Matrix([[1]]) - 1, Matrix([[1]]), domain='ZZ')

This looks weird but actually only the generator `x` has been replaced by `Matrix([[1]])`. The expression `Matrix([[1]])**3 + 2*Matrix([[1]]) - 1` is generated by the printer, it is not a part of the modified polynomial. Probably there should be some checking done when the generators of a polynomial are replaced.

**Comment by Jonty16117** [comment]
I'm working on this

**Comment by Jonty16117** [comment]
Hi, i tried it on sympy live, it gives the following result:
`Poly([1] + 2[1] - 1, [1], domain = Z)`

**Comment by Jonty16117** [comment]
> `from sympy import Poly`
> `from sympy.abc import x`
> `f = Poly(x**3 + 2*x - 1, x) `
> `f.subs(x, Matrix([1]))`
> 
> gives `Poly(Matrix([[1]])**3 + 2*Matrix([[1]]) - 1, Matrix([[1]]), domain='ZZ')`

can you tell me how to reproduce this particular result?

**Comment by sylee957** [comment]
The string output only depends on the printing
I think that you should use `pprint` to reproduce the original result

**Comment by Jonty16117** [comment]
> The string output only depends on the printing
> I think that you should use `pprint` to reproduce the original result

Indeed

**Comment by Saanidhyavats** [comment]
Not only in the case of Matrix but in this case as well the desired output is not coming:
`x=symbols('x')`
`f= Poly(x**3+2*x-1,x)`
` f.subs(x,[1])`
output:
`Poly(x**3 + 2*x - 1, x, domain='ZZ')`
Instead of putting `[1]` (list type) if I put `numpy.array(1)` then the things are working correctly.

**Comment by Jonty16117** [comment]
I am working on this

**Comment by Jonty16117** [comment]
I want to work on this issue. It would be great if anyone could guide me where to start from!

**Comment by sylee957** [comment]
In my opinion, I'd consider the substitutions like this an abuse, and should be considered a gotcha than an issue.
Maybe it can make use of documentation.

Perhaps it would be more better to substitute it with `MatrixSymbol` if you need matrix polynomial,
```
x = Symbol('x')
f = Poly(x**3 + 2*x - 1, x)
X = MatrixSymbol('X', 2, 2)
f.subs(x, X)
```

But it can be a loophole of bugs because this is treated as a scalar expression, than a matrix.

**Comment by sylee957** [comment]
And some other issues would remain if we introduce a matrix polynomial like that.
For example, how should the roots be computed?

**Comment by Saanidhyavats** [comment]
I think we can create a function for 1 dimensional matrix.

**Comment by sylee957** [comment]
But only supporting one dimensional matrix looks too trivial. I think that if some features are restricted, we should have multidimensional matrix polynomial as well.

**Comment by Saanidhyavats** [comment]
How will the function work for more than 1 dimensional matrix. For example in the above example instead of Matrix([[1]]) Matrix([2,3]) would be there then what is the expected result?

**Comment by sylee957** [comment]
Matrix polynomial can still make sense for square matrices.
Though for matrices larger than 1*1, you should consider that polynomials can be non-commutative.

**Comment by oscarbenjamin** [comment]
Does poly handle polynomials over non-commutative rings?

**Comment by sylee957** [comment]
It wouldn’t matter for scalar coefficients because whatever poly can handle should be able to be extended for matrix polynomial, but for generators,
 I think that it should only support univariate matrix polynomial if poly cannot handle noncommutativity.

**Comment by jksuom** [comment]
I'm unclear why it should be necessary to replace generators of a Poly object. There is an `eval` method for that kind of substitutions. I think that it could be made to work also with (square) matrices.

**Comment by sylee957** [comment]
One obvious limitation of `subs` is that scalar `1` isn't getting replaced to any matching identity matrix.

**Comment by jksuom** [comment]
I agree that something has to be done to fix that. One possibility might be writing `a*M**0` for a scalar `a` and matrix `M`.

**Comment by sylee957** [comment]
But one other reason I'd try to distinguish matrix polynomial is that evaluation can be done more efficiently, like Cayley-Hamilton theorem.

**Comment by jksuom** [comment]
Perhaps there could be a matrix method (or function) that takes a Poly object as a parameter.

**Comment by wsdookadr** [comment]
I wrote two [implementations here](https://stackoverflow.com/a/65678536/827519) related to evaluating a [Matrix polynomial](https://en.wikipedia.org/wiki/Matrix_polynomial) for a certain Matrix. One of the implementations supports multivariate polynomials too. Maybe a similar approach could work here.

**Comment by wsdookadr** [comment]
@oscarbenjamin 
> Does poly handle polynomials over non-commutative rings?

No

```python3
from sympy import *
x,y = symbols('x y', commutative=False)
p=Poly(x**2 * y)
[..]
PolynomialError: non-commutative expressions are not supported
```

**Comment by oscarbenjamin** [comment]
There are several types of matrix polynomial like `A*X**2 + B*X + C*I`

1. Scalar-matrix polynomial where coefficients `A, B, ...` are scalars but the "variable" `X` is a matrix.
2. Matrix-matrix polynomial where both coefficients `A, B, ...` are matrices and `X` is a matrix as well.
3. Matrix-scalar polynomial where coefficients `A, B, ...` are matrices but the "variable" `X` is a scalar.

The matrix-scalar polynomial is equivalent to a matrix whose elements are polynomials so that's already handled by the normal Matrix class.

The matrix-matrix polynomial can be efficiently evaluated by the Horner scheme.

A scalar-matrix polynomial can be evaluated more efficiently using the Paterson–Stockmeyer algorithm. The idea of this algorithm is to exchange matrix-matrix multiplies for scalar-matrix multiplies by grouping the terms. something like:
```
P(X) =
  A*X**9 + B*X**8 + C*X**7
+ D*X**6 + E*X**5 + F*X**4
+ G*X**3 + H*X**2 + J*X
+ K*I

P(X) =
  (A*X**3 + B*X**2 + C*X) * (X**3)**2
+ (D*X**3 + E*X**2 + F*X) * X**3
+ (G*X**3 + H*X**2 + J*X)
+ K*I

P2(X**3) =
  M3 * (X**3)**2
+ M2 * X**3
+ M1
+ K*I

```
This turns a scalar-matrix polynomial into a matrix-matrix polynomial of lower order in `X**3` which can then be computed with the Horner scheme. The number of matrix-matrix multiplies is optimal:

1. Compute `X**2 = X*X` (1 multiply)
2. Compute `X**3 = X**2 * X` (1 multiply)
3. Compute e.g. `M1 = G*X**3 + H*X**2 + J*X` using only scalar multiplication and addition
4. Using the Horner scheme `P2(X**3)` can be evaluated (2 multiplies)

Altogether that computes the matrix polynomial using 4 matrix-matrix multiplies instead of 9.

It would be useful to have a function for computing matrix-matrix polynomials and also scalar-matrix polynomials using the Paterson algorithm.

Another point to consider is that very high-order polynomials in large matrices can potentially be computed much more efficiently by first reducing the polynomial using the characteristic polynomial of the matrix.

Another possible optimisation is polynomial decomposition in the scalar-matrix case. Evaluating a matrix polynomial is expensive so it is probably worthwhile to attempt to compute a decomposition first.

Something else that might be worth having and that can be done more efficiently than evaluating a matrix polynomial is evaluating the product of a matrix polynomial with a column vector e.g.:
```
(A*X**2 + B*X + C*I) * v
```
This can be computed using only matrix-vector multiplies using the Horner scheme which can be much more efficient. This is how you could use the characteristic polynomial to solve a system of equations: the charpoly gives a polynomial form for the inverse of a matrix `M`. We don't actually need to compute `M**-1` though because we only want to know `(M**-1 v)` which can be computed without any matrix-matrix multiplies.
